### PR TITLE
LO: Better max reachable tier calc

### PR DESCRIPTION
**Pull request by robojumper**
Supersedes #9458. This does make LO a bit slower but if we're encouraging users to play with the tiers more in the new UI I think it's a price we need to pay to make the displayed info more accurate.

I did identify a bunch of speedups and simplifications since I last looked at this.

### CODE REVIEW

File: src/app/loadout-builder/process-worker/auto-stat-mod-utils.ts [hunk 1]
```diff
@@ -76,11 +76,17 @@ function doGeneralModsFit(
...
+  let generalModCosts = info.generalModCosts;
+  if (pickedMods.length) {
+    generalModCosts = generalModCosts.slice();
+    // Intentionally open-coded for performance
+    // eslint-disable-next-line @typescript-eslint/prefer-for-of
+    for (let i = 0; i < pickedMods.length; i++) {
+      generalModCosts.push(...pickedMods[i].generalModsCosts);
+    }
+    generalModCosts.sort((a, b) => b - a);
+  }
```
Comment (robojumper) [inline src/app/loadout-builder/process-worker/auto-stat-mod-utils.ts:89]: This is one of the hottest parts of the process and the previous approach with `flatMaps` and spreads was hard to optimize, it seems.

File: src/app/loadout-builder/filter/TierSelect.tsx [hunk 2]
```diff
@@ -77,7 +76,7 @@ export default function TierSelect({
                     {statRangesFiltered
                       ? t('LoadoutBuilder.MaxTier', {
                           tier: t('LoadoutBuilder.TierNumber', {
-                            tier: statTierWithHalf(statRangesFiltered[statHash].max),
+                            tier: statRangesFiltered[statHash].max,
```
Comment (robojumper) [inline src/app/loadout-builder/filter/TierSelect.tsx:79]: One small change is that this now returns full tiers only, not stat points (thus also not half tiers). But with the new UI only full reachable tiers are indicated anyway.

**Review by robojumper — COMMENTED** [review]
Some comments to make the changes easier to follow.

File: src/app/loadout-builder/process-worker/process-utils.ts [hunk 3]
```diff
@@ -136,7 +136,7 @@ export function updateMaxTiers(
     // Since we calculate the maximum tier we can hit for a stat in isolation,
     // require all other stats to hit their constrained minimums, but for this
     // stat we start from the highest tier we've observed.
-    const explorationStats = [...requiredMinimumExtraStats];
+    const explorationStats = requiredMinimumExtraStats.slice();
     explorationStats[statIndex] = minMax[statIndex].max * 10 - setStat;
```
Comment (bhollis) [inline src/app/loadout-builder/process-worker/process-utils.ts:140]: If you're testing performance, you might want to also try:

```
const explorationStats = requiredMinimumExtraStats.with(statIndex, minMax[statIndex].max * 10 - setStat)
```

File: src/app/loadout-builder/process-worker/process.ts [hunk 4]
```diff
@@ -233,6 +235,9 @@ export function process(
               const tier = tiers[index];
               const filter = resolvedStatConstraints[index];
               if (!filter.ignored) {
+                if (tier < statRangesFilteredInStatOrder[index].min) {
+                  statRangesFilteredInStatOrder[index].min = tier;
+                }
```
Comment (bhollis) [inline src/app/loadout-builder/process-worker/process.ts:240]: It's possible that this is a touch faster:
```suggestion
                const statRange = statRangesFilteredInStatOrder[index];
                if (tier < statRange.min) {
                  statRange.min = tier;
                }
```

File: src/app/loadout-builder/process-worker/process-utils.ts (hunk 3 above)
Comment (robojumper) [reply src/app/loadout-builder/process-worker/process-utils.ts:140]: Microbenchmarks suggest `with` is ever so slightly slower. I'd love to set up some benchmarks that yield tangible results (see #9993).

File: src/app/loadout-builder/process-worker/process-utils.ts [hunk 5]
```diff
@@ -49,6 +48,117 @@ export function precalculateStructures(
...
+      }
+      const neededValue = filter.minTier * 10 - value;
+      if (neededValue > 0) {
+        // All sets need at least these extra stats to hit minimums
+        requiredMinimumExtraStats[statIndex] = neededValue;
+      }
+    }
+  }
+
+  const numArtificeMods = items.filter((i) => i.isArtifice).length;
```
Comment (bhollis) [inline src/app/loadout-builder/process-worker/process-utils.ts:127]:
```suggestion
  const numArtificeMods = count((i) => i.isArtifice);
```

File: src/app/loadout-builder/process-worker/process-utils.ts [hunk 6]
```diff
@@ -49,6 +48,117 @@ export function precalculateStructures(
...
+
+  // Then, for every non-ignored stat where we haven't shown that we can hit T10...
+  for (let statIndex = 0; statIndex < statFiltersInStatOrder.length; statIndex++) {
+    const setStat = setStats[statIndex];
+    if (statFiltersInStatOrder[statIndex].ignored || minMax[statIndex].max >= 10) {
```
Comment (bhollis) [inline src/app/loadout-builder/process-worker/process-utils.ts:132]: Assigning `minMax[statIndex]` to a local could maybe help too

**Review by bhollis — APPROVED** [review]
Cool!

**Pull request merged**